acvm-repo/acvm/src/pwg/brillig.rs

```diff
@@ -14,28 +14,73 @@ use crate::{pwg::OpcodeNotSolvable, OpcodeResolutionError};
 
 use super::{get_value, insert_value};
 
-pub(super) struct BrilligSolver;
+pub(super) enum BrilligSolverStatus {
+    Finished,
+    InProgress,
+    ForeignCallWait(ForeignCallWaitInfo),
+}
 
-impl BrilligSolver {
-    pub(super) fn solve<B: BlackBoxFunctionSolver>(
-        initial_witness: &mut WitnessMap,
-        brillig: &Brillig,
-        bb_solver: &B,
+pub(super) struct BrilligSolver<'b, B: BlackBoxFunctionSolver> {
+    vm: VM<'b, B>,
+    acir_index: usize,
+}
+
+impl<'b, B: BlackBoxFunctionSolver> BrilligSolver<'b, B> {
+    /// Constructs a solver for a Brillig block given the bytecode and initial
+    /// witness. If the block should be skipped entirely because its predicate
+    /// evaluates to false, zero out the block outputs and return Ok(None).
+    pub(super) fn build_or_skip<'w>(
+        initial_witness: &'w mut WitnessMap,
+        brillig: &'w Brillig,
+        bb_solver: &'b B,
         acir_index: usize,
-    ) -> Result<Option<ForeignCallWaitInfo>, OpcodeResolutionError> {
+    ) -> Result<Option<Self>, OpcodeResolutionError> {
```

Comment:
I think that this function is trying to do too many things which is adding complexity into `solve_opcode`. We should probably handle the case where we're skipping the opcode separately.

Reply:
Agreed. I put everything together in an attempt to capture more of the logic to deal with Brillig blocks, but I agree it hurts readability.

Reply:
I have split `build_or_skip` as per your suggestion. Thanks!